Support mDNS responder probing

jukkar opened the pull request:
The mDNS probing is described in RFC 6762 chapter 8.1. The code will send an unsolicited mDNS query to network and will check if there are existing hosts with the same name. If there are, then the mDNS responder will not respond to queries it is configured to use.

Fixes #84333

jukkar committed:
net: dns: Introduce resolver function that avoids using the cache

Introduce dns_resolve_name_internal() that allows resolving a name
and not use DNS cache if caching is enabled.

This is needed in mDNS probing (RFC 6762 chapter 8.1) which needs
to send a mDNS query and not get any results from cache.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dns: Allow creating context with specific svc and port

Allow user to specify socket service struct and port number so
that DNS servers specified in Kconfig etc are not used. This way
we can send a DNS query to arbitrary address without it affecting
the system configuration.

This is used in mDNS probing so that that mDNS responder can send
probe message and not change / configure system global DNS config.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: mdns_responder: Enable probing in CI

Make sure we have a compile test in CI for probing support.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: mdns_responder: Implement probing support

The mDNS probing is described in RFC 6762 chapter 8.1.
The code will send an unsolicited mDNS query to network and
will check if there are existing hosts with the same name.
If there are, then the mDNS responder will not respond to
queries it is configured to use.

Fixes #84333

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -606,6 +616,268 @@ static int dns_read(int sock,
+				    uint32_t mgmt_event, struct net_if *iface)
+
+{
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
```
I'm not sure if connection manager events are a good choice here, note that the events are not tightly bound to a specific interface, I. e. if there are for example two interfaces in the system, one remains up and running, and the other one toggles up and down, there will be no event triggered. And if I understand correctly, the second interface should send a probe whenever it changes its state. Perhaps sticking to NET_IF_UP/DOWN events would be a better choice here?

jukkar replied:
It is just that if up/down is probably not good here because the network interface needs to have IP address before the probing can be done. Connection manager has these IPv6/IPv4 connected events, perhaps I could utilize those here.

jukkar replied:
After some initial testing, the `L4_IPVx_CONNECTED` is not good either as it is only triggered once. I will need to figure out something else for this, probably need to listen the IP address add signals and work from there.

rlubos replied:
Maybe monitor NET_IF_UP/DOWN **and** IP_ADDR_ADD/REMOVE events, but check `net_if_ipv4/6_get_global_addr()` and send probe only if it gives non-NULL result? Too bad we've dropped the idea of interface-specific connectivity events some time ago.

jukkar committed:
net: mdns_responder: Add announce support

After probing, send two unsolicited mDNS response messages
for our name.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos reviewed (commented):
Looks reasonalby so far, just one concern regarding connection manager use up for discussion.

jukkar commented:
* CI fixes
* Add announce support

jukkar commented:
* CI fixes
* Probe and announce fixes

The code is still missing some corner cases, will try to fix those.

jukkar committed:
net: dns: Allow creating context with specific interface

Allow user to specify the network interface where the query should
be sent, but only if the user has supplied the DNS servers in sockaddr
list. If user has specified a DNS servers in a string list, then the
possible network interface is taken from the string.

The reason for this is that it is not possible to add network interface
to the sockaddr list so an extra parameter is needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -720,6 +1200,72 @@ static int register_dispatcher(struct mdns_responder_context *ctx,
+	}
+#endif /* CONFIG_NET_IPV4 */
+
+	if (!ok) {
```
nit really, but aren't such bool checks on int against misra? Maybe we could just check `ok == 0 ` and return negative error here.

jukkar replied:
Indeed, will fix.

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -606,6 +650,442 @@ static int dns_read(int sock,
+					 &ctx->dispatcher.local_addr) && ret < PORT_COUNT);
+	if (ret >= PORT_COUNT) {
+		NET_ERR("No available port, %s probe fails!", "mDNS");
+		ret = -EIO;
```
We're mixing errno and DNS error space here, in error checks below we set `DNS_EAI_SYSTEM` as a return value. Shouldn't we set errno and return DNS_EAI_SYSTEM here too?

jukkar replied:
My bad, I mixed DNS and errno space. We should just use the returned value directly, no need to do DNS_EAI_SYSTEM here.

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -1481,21 +1472,275 @@ static int init_listener(void)
+		net_buf_add_be32(answer, ttl);
+
+		if (type == DNS_RR_TYPE_A) {
+			net_buf_add_be16(answer, 4);
```
Maybe use `sizeof(struct in_addr)`/`sizeof(struct in6_addr)` instead of magic numbers?

jukkar replied:
Yes, will do.

rlubos reviewed (commented):
Complex stuff! Pretty neat though.

jukkar commented:
* Rebased against latest main
* Resolved the merge conflict

jukkar commented:
To all, Friday is a deadline for this PR to get to 4.1 release, please take a look if you spot anything obviously wrong in the code, or try it yourself in real network. The PR became somewhat complex, because of the multiple network interface support, for this rather simple feature.

pdgendt reviewed (approved):
Don't have time for testing though.
